mars/session.py

```diff
@@ -101,58 +105,48 @@ def run(self, *tensors, **kw):
             ret_list = True
 
         tensors = tuple(mt.tensor(t) for t in tensors)
-        results = [None] * len(tensors)
-        idx_to_run_tensors = OrderedDict()
-        idx_to_fetch_tensors = OrderedDict()
-
-        # those executed tensors should fetch data directly, submit the others
-        for i, t in enumerate(tensors):
-            if t.key in self._executed_keys:
-                idx_to_fetch_tensors[i] = t
-            else:
-                idx_to_run_tensors[i] = t
-
-        # execute the non-executed tensors
-        if idx_to_run_tensors:
-            execute_result = self._sess.run(*idx_to_run_tensors.values(), **kw)
-            if execute_result:
-                # fetch is True
-                for j, result in zip(idx_to_run_tensors, execute_result):
-                    results[j] = result
-            run_tensors = list(idx_to_run_tensors.values())
-            self._executed_keys.update(t.key for t in run_tensors)
-            for t in run_tensors:
-                t._execute_session = self
+        result = self._sess.run(*tensors, **kw)
+
+        for t in tensors:
+            t._execute_session = self
```

Comment:
Shall we consider that a tensor could be executed by different sessions?

Reply:
Yeah, and do decref for all sessions if the tensor is garbage collected